### PR TITLE
Pandora shard info into beacon block body

### PR DESCRIPTION
**Pull request by atif-konasl**
**What type of PR is this?**

> Feature

**What does this PR do? Why is it needed?**

According to lukso mainnet architecture, pandora chain acts like an executable sharding chain so we need to maintain sharding info into beacon block body in vanguard client.

**Which issues(s) does this PR fix?**
issue no: [3](https://github.com/lukso-network/vanguard-consensus-engine/issues/3)

**Other notes for review**

### CODE REVIEW

File: beacon-chain/blockchain/process_block.go [hunk 1]
```diff
@@ -105,6 +105,13 @@ func (s *Service) onBlock(ctx context.Context, signed *ethpb.SignedBeaconBlock,
 	if !valid {
 		return errors.New("signature in block failed to verify")
 	}
+	// Vanguard: Validated by vanguard node. Now intercepting the execution and publishing the block
+	// and waiting for confirmation from orchestrator. If Lukso vanguard flag is enabled then these segment of code will be executed
+	if s.enableVanguardNode {
```
Comment (blazejkrzak) [inline beacon-chain/blockchain/process_block.go:110]: do not nest ifs

File: beacon-chain/blockchain/process_block.go [hunk 2]
```diff
@@ -245,6 +252,15 @@ func (s *Service) onBlockBatch(ctx context.Context, blks []*ethpb.SignedBeaconBl
 	if !verify {
 		return nil, nil, errors.New("batch block signature verification failed")
 	}
+	// Vanguard: Validated by vanguard node. Now intercepting the execution and publishing the block
+	// and waiting for confirmation from orchestrator. If Lukso vanguard flag is enabled then these segment of code will be executed
+	if s.enableVanguardNode {
```
Comment (blazejkrzak) [inline beacon-chain/blockchain/process_block.go:257]: use strategy pattern instead of nesting ifs

File: beacon-chain/blockchain/van_process_pending_blocks.go [hunk 3]
```diff
@@ -24,6 +25,7 @@ var (
 	errPendingBlockTryLimitExceed = errors.New("maximum wait is exceeded and orchestrator can not verify the block")
 	errUnknownStatus              = errors.New("invalid status from orchestrator")
 	errInvalidRPCClient           = errors.New("invalid orchestrator rpc client or no client initiated")
+	errPendingQueueUnprocessed    = errors.New("pending queue is un-processed")
```
Comment (blazejkrzak) [inline beacon-chain/blockchain/van_process_pending_blocks.go:28]: typeo here

File: beacon-chain/blockchain/van_process_pending_blocks.go [hunk 4]
```diff
@@ -32,7 +34,38 @@ type PendingBlocksFetcher interface {
...
 }
 
-type blockRoot [32]byte
+// BlockProposal interface use when validator calls GetBlock api for proposing new beancon block
+type PendingQueueFetcher interface {
+	CanPropose() error
+}
+
+// CanPropose
+func (s *Service) CanPropose() error {
```
Comment (blazejkrzak) [inline beacon-chain/blockchain/van_process_pending_blocks.go:43]: where is a test for this?

File: beacon-chain/blockchain/van_process_pending_blocks.go [hunk 5]
```diff
@@ -32,7 +34,38 @@ type PendingBlocksFetcher interface {
...
+
+	if len(blks) > 0 {
+		log.WithField("unprocessedBlockLen", len(blks)).WithError(err).Error("Pending queue is not nil")
+		return errPendingQueueUnprocessed
+	}
+	return nil
+}
+
+// UnConfirmedBlocksFromCache retrieves all the cached blocks from cache and send it back to event api
+func (s *Service) SortedUnConfirmedBlocksFromCache() ([]*ethpb.BeaconBlock, error) {
```
Comment (blazejkrzak) [inline beacon-chain/blockchain/van_process_pending_blocks.go:57]: where is the test for this?

File: beacon-chain/core/helpers/committee_test.go [hunk 6]
```diff
@@ -660,20 +660,20 @@ func TestPrecomputeProposerIndices_Ok(t *testing.T) {
...
-	state, err := beaconstate.InitializeFromProto(&pb.BeaconState{
-		Validators:  validators,
-		RandaoMixes: make([][]byte, params.BeaconConfig().EpochsPerHistoricalVector),
-	})
-	require.NoError(t, err)
-	proposerAssignments, err = ProposerAssignments(state, types.Epoch(0))
-	require.NoError(t, err)
-	assert.Equal(t, 31, len(proposerAssignments))
-}
+//func TestProposerAssignments(t *testing.T) {
```
Comment (blazejkrzak) [inline beacon-chain/core/helpers/committee_test.go:663]: do not leave commented code

Comment (blazejkrzak) [reply beacon-chain/core/helpers/committee_test.go:663]: What had happen here? Why you commented test code? What kind of sorcery is this?

File: beacon-chain/node/node.go [hunk 7]
```diff
@@ -674,8 +676,10 @@ func (b *BeaconNode) registerRPCService() error {
 		EnableDebugRPCEndpoints: enableDebugRPCEndpoints,
 		MaxMsgSize:              maxMsgSize,
 
-		// Vanguard: un-confirmed cached block fetcher
+		// vanguard: EnableVanguardNode and UnconfirmedBlockFetcher is used for vanguard chain
+		EnableVanguardNode:      b.cliCtx.Bool(cmd.VanguardNetwork.Name),
```
Comment (blazejkrzak) [inline beacon-chain/node/node.go:680]: I'd suggest to rename it. IMHO vanguard is more like a mode, so maybe `VanguardModeEnabled`? Also I'd suggest to use past tense in such flags

File: beacon-chain/rpc/service.go [hunk 8]
```diff
@@ -69,6 +69,7 @@ type Service struct {
 	chainStartFetcher       powchain.ChainStartFetcher
 	mockEth1Votes           bool
 	enableDebugRPCEndpoints bool
+	enableVanguardNode      bool // vanguard: vanguard chain enable flag
```
Comment (blazejkrzak) [inline beacon-chain/rpc/service.go:72]: same here. In my opinion name is inaccurate

File: beacon-chain/rpc/service.go [hunk 9]
```diff
@@ -124,6 +126,7 @@ type Config struct {
 	GenesisFetcher          blockchain.GenesisFetcher
 	EnableDebugRPCEndpoints bool
 	MockEth1Votes           bool
+	EnableVanguardNode      bool // vanguard: vanguard chain enable flag
```
Comment (blazejkrzak) [inline beacon-chain/rpc/service.go:129]: same here. In my opinion name is inaccurate

File: beacon-chain/rpc/service.go [hunk 10]
```diff
@@ -191,7 +195,9 @@ func NewService(ctx context.Context, cfg *Config) *Service {
 		maxMsgSize:              cfg.MaxMsgSize,
 
 		// Vanguard: un-confirmed cached block fetcher
+		enableVanguardNode:      cfg.EnableVanguardNode,
```
Comment (blazejkrzak) [inline beacon-chain/rpc/service.go:198]: same here. In my opinion name is inaccurate

File: beacon-chain/rpc/service.go [hunk 11]
```diff
@@ -265,6 +271,10 @@ func (s *Service) Start() {
 		PendingDepositsFetcher: s.pendingDepositFetcher,
 		SlashingsPool:          s.slashingsPool,
 		StateGen:               s.stateGen,
+
+		// vanguard: initiate pending queue fetcher
+		EnableVanguardNode:  s.enableVanguardNode,
```
Comment (blazejkrzak) [inline beacon-chain/rpc/service.go:276]: same here. In my opinion name is inaccurate

File: beacon-chain/rpc/validator/proposer.go [hunk 12]
```diff
@@ -59,6 +59,16 @@ func (vs *Server) GetBlock(ctx context.Context, req *ethpb.BlockRequest) (*ethpb
 		return nil, status.Errorf(codes.Unavailable, "Syncing to latest head, not ready to respond")
 	}
 
+	// vanguard: If vanguard chain flag is enabled then we need to check the length of pending queue size. if
+	// pending queue is not empty that means syncing and verification process does not complete yet so skipped the
+	// slot
+	if vs.EnableVanguardNode {
```
Comment (blazejkrzak) [inline beacon-chain/rpc/validator/proposer.go:65]: do not nest ifs

File: beacon-chain/rpc/validator/proposer.go [hunk 13]
```diff
@@ -135,6 +145,24 @@ func (vs *Server) GetBlock(ctx context.Context, req *ethpb.BlockRequest) (*ethpb
 	}
 	blk.StateRoot = stateRoot
 
+	// If vanguard chain is enabled, we set the latest pandora sharding info into beacon block so that
+	// we do not need to make another api call for getting latest sharding info
+	if vs.EnableVanguardNode {
```
Comment (blazejkrzak) [inline beacon-chain/rpc/validator/proposer.go:150]: do not nest ifs

File: beacon-chain/rpc/validator/server.go [hunk 14]
```diff
@@ -59,10 +59,14 @@ type Server struct {
 	ExitPool               voluntaryexits.PoolManager
 	BlockReceiver          blockchain.BlockReceiver
 	MockEth1Votes          bool
+	EnableVanguardNode     bool // vanguard: vanguard chain enable flag
```
Comment (blazejkrzak) [inline beacon-chain/rpc/validator/server.go:62]: same here. In my opinion name is inaccurate

File: beacon-chain/rpc/validator/van_proposer_test.go [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+	require.NoError(t, err)
+
+	graffiti := bytesutil.ToBytes32([]byte("eth2"))
+	req := &ethpb.BlockRequest{
+		Slot:         1,
+		RandaoReveal: randaoReveal,
+		Graffiti:     graffiti[:],
+	}
+
+	proposerSlashings := make([]*ethpb.ProposerSlashing, params.BeaconConfig().MaxProposerSlashings)
```
Comment (blazejkrzak) [inline beacon-chain/rpc/validator/van_proposer_test.go:70]: Is there any reason to process slashing in this test?

File: beacon-chain/rpc/validator/van_proposer_test.go [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+	require.NoError(t, err)
+	prevStateRoot := block.StateRoot
+
+	pandoraHeader, _ := testutil.NewPandoraBlock(block.Slot, 120)
+	newBeaconBlock := testutil.NewBeaconBlockWithPandoraSharding(pandoraHeader, block.Slot)
+	block.Body.PandoraShard = newBeaconBlock.Block.Body.PandoraShard
+
+	block, err = proposerServer.UpdateStateRoot(ctx, block)
+	require.NoError(t, err)
+	assert.DeepNotEqual(t, prevStateRoot, block.StateRoot)
```
Comment (blazejkrzak) [inline beacon-chain/rpc/validator/van_proposer_test.go:105]: Is there any better way to determine that state is correct? DeepNotEqual seems not suitable for testing the side effect. You just only check that state is different, but you do not have desired state.

File: beacon-chain/state/cloners.go [hunk 17]
```diff
@@ -284,3 +285,24 @@ func CopyValidator(val *ethpb.Validator) *ethpb.Validator {
 		WithdrawableEpoch:          val.WithdrawableEpoch,
 	}
 }
+
+// CopyPandoraShard copies PandoraShard
+func CopyPandoraShard(pShards []*ethpb.PandoraShard) []*ethpb.PandoraShard {
```
Comment (blazejkrzak) [inline beacon-chain/state/cloners.go:290]: please use full names instead of abbreviations

File: deps.bzl [hunk 18]
```diff
@@ -729,8 +729,9 @@ def prysm_deps():
         name = "com_github_ferranbt_fastssz",
         importpath = "github.com/ferranbt/fastssz",
         nofuzz = True,
-        sum = "h1:9VDpsWq096+oGMDTT/SgBD/VgZYf4pTF+KTPmZ+OaKM=",
-        version = "v0.0.0-20210120143747-11b9eff30ea9",
+        replace = "github.com/atif-konasl/fastssz",
```
Comment (blazejkrzak) [inline deps.bzl:732]: What is the motivation of this change?
IMHO we should not change it to your repository. Push it to lukso namespace or rethink if we need this replacement at all

File: go.mod [hunk 19]
```diff
@@ -4,8 +4,9 @@ go 1.16
 
 replace (
 	github.com/ethereum/go-ethereum => github.com/prysmaticlabs/bazel-go-ethereum v0.0.0-20210222122116-71d15f72c132
+	github.com/ferranbt/fastssz => github.com/atif-konasl/fastssz v0.0.0-20210705113036-087ec0cbb160
```
Comment (blazejkrzak) [inline go.mod:7]: What is the motivation of this change?
IMHO we should not change it to your repository. Push it to lukso namespace or rethink if we need this replacement at all

File: go.sum [hunk 20]
```diff
@@ -100,6 +100,8 @@ github.com/armon/consul-api v0.0.0-20180202201655-eb2c6b5be1b6/go.mod h1:grANhF5
 github.com/armon/go-metrics v0.0.0-20180917152333-f0300d1749da/go.mod h1:Q73ZrmVTwzkszR9V5SSuryQ31EELlFMUz1kKyl939pY=
 github.com/armon/go-radix v0.0.0-20180808171621-7fddfc383310/go.mod h1:ufUuZ+zHj4x4TnLV4JWEpy2hxWSpsRywHrMgIH9cCH8=
 github.com/aryann/difflib v0.0.0-20170710044230-e206f873d14a/go.mod h1:DAHtR1m6lCRdSC2Tm3DSWRPvIPr6xNKyeHdqDQSQT+A=
+github.com/atif-konasl/fastssz v0.0.0-20210705113036-087ec0cbb160 h1:VxX7vk/C1ouWFMXfF078nZ1WMZEbzjwMxpc/hHHyluY=
```
Comment (blazejkrzak) [inline go.sum:103]: What is the motivation of this change?
IMHO we should not change it to your repository. Push it to lukso namespace or rethink if we need this replacement at all

File: shared/testutil/block.go [hunk 21]
```diff
@@ -569,30 +541,26 @@ func NewPandoraBlock(
...
-		header.GasUsed,
-		header.Time,
-		header.Extra,
-	}); err != nil {
-		return eth1Types.EmptyRootHash
-	}
-	hasher.Sum(hash[:0])
-	return hash
+// NewBeaconBlockWithPandoraSharding
+func NewBeaconBlockWithPandoraSharding(panHeader *gethTypes.Header, slot types.Slot) *ethpb.SignedBeaconBlock {
```
Comment (blazejkrzak) [inline shared/testutil/block.go:548]: please rename it to `NewBeaconBlockWithPandoraShard` 
Sharding is a verb, Shard is a noun. You should use noun because there is no action in function block

File: validator/client/pan_propose.go [hunk 22]
```diff
@@ -0,0 +1,274 @@
...
+// - Signs header hash using a validator key
+// - Submit signature and header to pandora node
+func (v *validator) processPandoraShardHeader(
+	ctx context.Context,
+	beaconBlk *ethpb.BeaconBlock,
+	slot types.Slot,
+	epoch types.Epoch,
+	pubKey [48]byte,
+) error {
+
```
Comment (blazejkrzak) [inline validator/client/pan_propose.go:52]: Redundant newline

File: validator/client/pan_propose.go [hunk 23]
```diff
@@ -0,0 +1,274 @@
...
+	// if pandoraShard is nil means there is no pandora blocks in pandora chain except block-0
+	if beaconBlk.Body != nil && beaconBlk.Body.PandoraShard != nil {
+		latestPandoraHash = common.BytesToHash(beaconBlk.Body.PandoraShard[0].Hash)
+		latestPandoraBlkNum = beaconBlk.Body.PandoraShard[0].BlockNumber
+	}
+
+	// Request for pandora chain header
+	header, headerHash, extraData, err := v.pandoraService.GetShardBlockHeader(ctx, latestPandoraHash, latestPandoraBlkNum+1)
+	if err != nil {
```
Comment (blazejkrzak) [inline validator/client/pan_propose.go:65]: please do not nest ifs

File: validator/client/pan_propose.go [hunk 24]
```diff
@@ -0,0 +1,274 @@
...
+			WithField("fmtKey", fmtKey).
+			WithError(err).Error("Failed to request block header from pandora node")
+		if v.emitAccountMetrics {
+			ValidatorProposeFailVec.WithLabelValues(fmtKey).Inc()
+		}
+		return err
+	}
+	// TODO- Need to verify pandora parentHash
+	// Validate pandora chain header hash, extraData fields
+	if err := v.verifyPandoraShardHeader(slot, epoch, header, headerHash, extraData); err != nil {
```
Comment (blazejkrzak) [inline validator/client/pan_propose.go:76]: please do not nest ifs. Use early return statements

File: validator/client/pan_propose.go [hunk 25]
```diff
@@ -0,0 +1,274 @@
...
+		return err
+	}
+	headerHashSig, err := v.keyManager.Sign(ctx, &validatorpb.SignRequest{
+		PublicKey:       pubKey[:],
+		SigningRoot:     headerHash[:],
+		SignatureDomain: nil,
+		Object:          nil,
+	})
+
+	if err != nil {
```
Comment (blazejkrzak) [inline validator/client/pan_propose.go:92]: please do not nest ifs

File: validator/client/pan_propose.go [hunk 26]
```diff
@@ -0,0 +1,274 @@
...
+	latestPandoraBlkNum := uint64(0)
+
+	// if pandoraShard is nil means there is no pandora blocks in pandora chain except block-0
+	if beaconBlk.Body != nil && beaconBlk.Body.PandoraShard != nil {
+		latestPandoraHash = common.BytesToHash(beaconBlk.Body.PandoraShard[0].Hash)
+		latestPandoraBlkNum = beaconBlk.Body.PandoraShard[0].BlockNumber
+	}
+
+	// Request for pandora chain header
+	header, headerHash, extraData, err := v.pandoraService.GetShardBlockHeader(ctx, latestPandoraHash, latestPandoraBlkNum+1)
```
Comment (blazejkrzak) [inline validator/client/pan_propose.go:64]: I think that you should restrict yourself to the `highestPandoraHash` not `latestPandoraHash`

File: validator/client/pan_propose.go [hunk 27]
```diff
@@ -0,0 +1,274 @@
...
+		}
+		return err
+	}
+
+	header.MixDigest = common.BytesToHash(headerHashSig.Marshal())
+	var headerHashSig96Bytes [96]byte
+	copy(headerHashSig96Bytes[:], headerHashSig.Marshal())
+
+	// Submit bls signature to pandora
+	if status, err := v.pandoraService.SubmitShardBlockHeader(
```
Comment (blazejkrzak) [inline validator/client/pan_propose.go:104]: Do not nest ifs

File: validator/client/pan_propose.go [hunk 28]
```diff
@@ -0,0 +1,274 @@
...
+
+	return nil
+}
+
+// panShardingCanonicalInfo method gets header hash and block number from sharding head
+func (v *validator) updateStateRoot(
+	ctx context.Context,
+	beaconBlk *ethpb.BeaconBlock,
+) (*ethpb.BeaconBlock, error) {
+
```
Comment (blazejkrzak) [inline validator/client/pan_propose.go:228]: newline

File: validator/client/pan_propose.go [hunk 29]
```diff
@@ -0,0 +1,274 @@
...
+// verifyPandoraShardHeader verifies pandora sharding chain header hash and extraData field
+func (v *validator) verifyPandoraShardHeader(
+	slot types.Slot,
+	epoch types.Epoch,
+	header *eth1Types.Header,
+	headerHash common.Hash,
+	extraData *pandora.ExtraData,
+) error {
+
```
Comment (blazejkrzak) [inline validator/client/pan_propose.go:165]: newline

File: validator/client/pan_propose_test.go [hunk 30]
```diff
@@ -0,0 +1,152 @@
...
+// TestVerifyPandoraHeader_Ok method checks pandora header validation method
+func TestVerifyPandoraShardHeader(t *testing.T) {
+	validator, _, _, finish := setup(t)
+	validator.enableVanguardNode = true
+	defer finish()
+
+	blk := testutil.NewBeaconBlock()
+	blk.Block.Slot = 98
+	blk.Block.ProposerIndex = 23
+	epoch := types.Epoch(uint64(blk.Block.Slot) / 32)
```
Comment (blazejkrzak) [inline validator/client/pan_propose_test.go:30]: is there any reason to hardcode 32 instead of using params?

File: shared/testutil/block.go [hunk 31]
```diff
@@ -511,36 +511,8 @@ func HydrateV1BeaconBlockBody(b *v1.BeaconBlockBody) *v1.BeaconBlockBody {
...
-
-	return
-}
-
 // getDummyBlock method creates a brand new block with extraData
-func NewPandoraBlock(
-	slot types.Slot,
-	proposerIndex uint64,
-) (*gethTypes.Header, common.Hash, *pandora.ExtraData) {
+func NewPandoraBlock(slot types.Slot, proposerIndex uint64) (*gethTypes.Header, *pandora.ExtraData) {
```
Comment (blazejkrzak) [inline shared/testutil/block.go:515]: why you use proposerIndex?

File: validator/client/propose.go [hunk 32]
```diff
@@ -99,16 +80,13 @@ func (v *validator) ProposeBlock(ctx context.Context, slot types.Slot, pubKey [4
...
-	// processPandoraShardHeader method process the block header from pandora chain
-	if status, err := v.processPandoraShardHeader(ctx, b, slot, epoch, pubKey); !status || err != nil {
-		log.WithError(err).
-			WithField("pubKey", fmt.Sprintf("%#x", pubKey)).
-			WithField("slot", slot).
-			Error("Failed to process pandora chain shard header")
-		if v.emitAccountMetrics {
-			ValidatorProposeFailVec.WithLabelValues(fmtKey).Inc()
+	// Vanguard: If VanguardNetwork flag is enabled then this if state will be executed
+	if v.enableVanguardNode {
```
Comment (blazejkrzak) [inline validator/client/propose.go:84]: do not nest ifs

File: validator/client/pan_propose_test.go [hunk 33]
```diff
@@ -0,0 +1,152 @@
...
+	validator, m, _, finish := setup(t)
+	validator.enableVanguardNode = true
+	defer finish()
+
+	secretKey, err := bls.SecretKeyFromBytes(bytesutil.PadTo([]byte{1}, 32))
+	require.NoError(t, err, "Failed to generate key from bytes")
+	publicKey := secretKey.PublicKey()
+	var pubKey [48]byte
+	copy(pubKey[:], publicKey.Marshal())
+	km := &mockKeymanager{
```
Comment (blazejkrzak) [inline validator/client/pan_propose_test.go:71]: try to use full names

File: beacon-chain/blockchain/van_process_pending_blocks.go [hunk 34]
```diff
@@ -32,7 +34,38 @@ type PendingBlocksFetcher interface {
 	SortedUnConfirmedBlocksFromCache() ([]*ethpb.BeaconBlock, error)
 }
 
-type blockRoot [32]byte
+// BlockProposal interface use when validator calls GetBlock api for proposing new beancon block
```
Comment (mxmar) [inline beacon-chain/blockchain/van_process_pending_blocks.go:37]: `typo` :) It should be `beacon` here.

File: validator/client/propose.go [hunk 35]
```diff
@@ -25,24 +18,12 @@ import (
...
-	errInvalidProposerIndex = errors.New("invalid proposer index")
-	// errInvalidTimestamp is returned if the timestamp of a block is higher than the current time
-	errInvalidTimestamp = errors.New("invalid timestamp")
+	"time"
 )
 
 type signingFunc func(context.Context, *validatorpb.SignRequest) (bls.Signature, error)
 
-const domainDataErr = "could not getverify domain data"
+const domainDataErr = "could not get verify domain data"
```
Comment (blazejkrzak) [inline validator/client/propose.go:26]: `could not get verified domain data`
or
`could not verify domain data`

File: validator/client/propose_test.go [hunk 36]
```diff
@@ -236,23 +227,6 @@ func TestProposeBlock_BlocksDoubleProposal(t *testing.T) {
 		gomock.Any(), // epoch
 	).Times(3).Return(&ethpb.DomainResponse{SignatureDomain: make([]byte, 32)}, nil /*err*/)
 
-	header, headerHash, extraData := testutil.NewPandoraBlock(slot, 0)
```
Comment (blazejkrzak) [inline validator/client/propose_test.go:239]: why is it removed?

File: validator/client/propose_test.go [hunk 37]
```diff
@@ -358,18 +315,6 @@ func TestProposeBlock_AllowsPastProposals(t *testing.T) {
 		gomock.Any(),
 	).Return(blk.Block, nil /*err*/)
 
-	header, headerHash, extraData := testutil.NewPandoraBlock(farAhead, 0)
```
Comment (blazejkrzak) [inline validator/client/propose_test.go:361]: why is it removed?

File: validator/pandora/client.go [hunk 38]
```diff
@@ -66,9 +66,16 @@ func (oc *PandoraClient) Close() error {
 //  - result[1], 32 bytes hex encoded receipt hash for transaction proof
 //  - result[2], hex encoded rlp block header
 //  - result[3], hex encoded block number
-func (oc *PandoraClient) GetShardBlockHeader(ctx context.Context) (*ShardBlockHeaderResponse, error) {
+func (oc *PandoraClient) GetShardBlockHeader(
+	ctx context.Context,
+	parentHash common.Hash,
+	nextBlockNumber uint64,
+) (*ShardBlockHeaderResponse, error) {
+
```
Comment (blazejkrzak) [inline validator/pandora/client.go:74]: newline

File: beacon-chain/core/helpers/committee_test.go (hunk 6 above)
Comment (mxmar) [reply beacon-chain/core/helpers/committee_test.go:663]: This behavior can create regression

File: validator/pandora/service_test.go [hunk 39]
```diff
@@ -68,7 +70,7 @@ func TestService_GetShardBlockHeader_Success(t *testing.T) {
 	pandoraService.connected = true
 	pandoraService.isRunning = true
 
-	actualHeader, actualHash, actualExtraData, err := pandoraService.GetShardBlockHeader(context.Background())
+	actualHeader, actualHash, actualExtraData, err := pandoraService.GetShardBlockHeader(context.Background(), types.EmptyRootHash, 1000)
```
Comment (blazejkrzak) [inline validator/pandora/service_test.go:73]: It would be great to have failing scenarios covered by test

**Review by blazejkrzak — CHANGES_REQUESTED** [review]
Things to change:
- codestyle
- typeos
- write missing tests
- remove unecessary comments
- uncomment failing tests. They are for a purpose. Solution is to fix them not comment them.
- Fix tests not remove them. If they are redundant now, why they were present in a first place? If something is really redundant please describe it in comment in this pr because I may miss some scope since github overview is far inferior than IDE

**Comment by blazejkrzak** [comment]
@atif-konasl Where are the implementation notes?

File: deps.bzl (hunk 18 above)
Comment (mxmar) [reply deps.bzl:732]: Agree, `lukso-network` namespace required

File: go.mod (hunk 19 above)
Comment (mxmar) [reply go.mod:7]: Agree, lukso-network namespace required

File: shared/testutil/block.go (hunk 31 above)
Comment (mxmar) [reply shared/testutil/block.go:515]: @atif-konasl I think there is a `typo` also: `getDummyBlock` vs `NewPandoraBlock`

File: validator/node/node.go [hunk 40]
```diff
@@ -434,7 +434,9 @@ func (c *ValidatorClient) registerValidatorService(
 		WalletInitializedFeed:      c.walletInitialized,
 		GraffitiStruct:             gStruct,
 		LogDutyCountDown:           c.cliCtx.Bool(flags.EnableDutyCountDown.Name),
-		PandoraService:             pandoraService,
+		// Vanguard: pandora service and vanguard node flag are needed for vanguard chain
+		PandoraService:     pandoraService,
+		EnableVanguardNode: c.cliCtx.Bool(cmd.VanguardNetwork.Name),
```
Comment (mxmar) [inline validator/node/node.go:439]: As @blazejkrzak has mentioned I would prefer use for example `VanguardModeEnabled`

File: validator/client/pan_propose.go [hunk 41]
```diff
@@ -0,0 +1,274 @@
...
+	if nil != err {
+		return err
+	}
+
+	// verify slot number
+	if extraData.Slot != uint64(slot) {
+		log.WithError(errInvalidSlot).
+			WithField("slot", slot).
+			WithField("extraDataSlot", extraData.Slot).
+			WithField("header", header.Extra).
```
Comment (mxmar) [inline validator/client/pan_propose.go:193]: Maybe lets name it `WithField("headerExtra", header.Extra).`

File: validator/client/pan_propose.go [hunk 42]
```diff
@@ -0,0 +1,274 @@
...
+		v.genesisTime,
+		types.Slot(extraData.Slot),
+		params.BeaconNetworkConfig().MaximumGossipClockDisparity,
+	)
+
+	if nil != err {
+		log.WithError(errInvalidSlot).
+			WithField("slot", slot).
+			WithField("extraDataSlot", extraData.Slot).
+			WithField("header", header.Extra).
```
Comment (mxmar) [inline validator/client/pan_propose.go:210]: same here - I will suggest `WithField("headerExtra", header.Extra).`

File: validator/client/pan_propose.go [hunk 43]
```diff
@@ -0,0 +1,274 @@
...
+		header.Difficulty,
+		header.Number,
+		header.GasLimit,
+		header.GasUsed,
+		header.Time,
+		header.Extra,
+	}); err != nil {
+		return eth1Types.EmptyRootHash
+	}
+	hasher.Sum(hash[:0])
```
Comment (mxmar) [inline validator/client/pan_propose.go:284]: I think it's OK, but more readable is to use `hasher.Sum(hash[nil])`

**Review by mxmar — CHANGES_REQUESTED** [review]
@atif-konasl A lot of implementation has been done there. Good work 👍 

Please take a look on my comments :)
